### DISCUSSION TITLE
How to avoid user routing, refreshing or killing the tab

### DISCUSSION THREAD

**Discussion by prafulrajhm** [vercel/next.js#21896] [Help]
Hi all, I have a scenario where the user needs to submit the form before leaving the page once.

How to handle scenarios like refreshing the page or killing tab/browser. I tried this below logic which works fine for the next routings (Links and router routes) but not refreshing the page or killing tab/browser.
`useEffect(() => {
    const routeChangeStart = url => {
      if (Router.asPath !== url && !quickTestData.submit) {
        setAskConfirm(true)
        Router.events.emit('routeChangeError')
        Router.replace(Router, path)
        throw 'Abort route change. Please ignore this error.'
      }
    }

    const beforeunload = e => {
      console.log('unload detect')
      if (!quickTestData.submit) {
        e.preventDefault()
        setTimeOut(true)
      }
    }

    window.addEventListener('beforeunload', beforeunload)
    Router.events.on('routeChangeStart', routeChangeStart)

    window.onpopstate = () => {
      setAskConfirm(true)
    }

    return () => {
      window.removeEventListener('onbeforeunload', beforeunload)
      Router.events.off('routeChangeStart', routeChangeStart)
      console.log('unmounting')
    }
  }, [])`

**Comment by Noitidart** [comment]
Did you find a solution?
